apps/prairielearn/src/question-servers/freeform.ts

```diff
@@ -1756,29 +1873,33 @@ export async function parse(submission, variant, question, course) {
   });
 }
 
-export async function grade(submission, variant, question, question_course) {
+export async function grade(
+  submission: Submission,
+  variant: Variant,
+  question: Question,
+  question_course: Course,
+): QuestionServerReturnValue<Partial<GradeResultData>> {
   return instrumented('freeform.grade', async () => {
     debug('grade()');
     if (variant.broken_at) throw new Error('attempted to grade broken variant');
     if (submission.broken) throw new Error('attempted to grade broken submission');
 
     const context = await getContext(question, question_course);
-    let data = {
+    const data = {
       // Note that `params` and `true_answer` can change during `parse()`, so we
       // use the submission's values when grading.
-      params: submission.params,
-      correct_answers: submission.true_answer,
-      submitted_answers: submission.submitted_answer,
-      format_errors: submission.format_errors,
+      params: submission.params ?? {},
+      correct_answers: submission.true_answer ?? {},
+      submitted_answers: submission.submitted_answer ?? {},
+      format_errors: submission.format_errors ?? {},
```

Comment:
These and the below changes should be moved to another PR for independent discussion.

Reply:
Moved to https://github.com/PrairieLearn/PrairieLearn/pull/11832.